Replace UnifiedViewingKey with UnifiedFullViewingKey

str4d opened the pull request:
UnifiedViewingKey had a bug that made it incompatible with ZIP 316.

For compatibility with the current `zcash/librustzcash` revision we use
a temporary fake UFVK encoding that stores sufficient information to use
the current APIs, and a superset of the actual ZIP 316 FVK information.

Depends on https://github.com/zcash/librustzcash/pull/570.

Closes zcash/zcash-android-wallet-sdk#525.
Closes zcash/zcash-android-wallet-sdk#526.

# Author
<!-- NOTE: Do not modify these when initially opening the pull request.  This is a checklist template that you tick off AFTER the pull request is created. -->
- [ ] Self-review: Did you review your own code in GitHub's web interface? _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Manual tests: Did you update the [manual tests](../blob/main/docs/tests) as appropriate? _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
- [x] Documentation: Did you update documentation as appropriate? (e.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the demo app and try the changes?
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._


# Reviewer

- [x] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) checklist?
- [x] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after changes and when possible, leave it in a better place._
- [x] Documentation: Did you review Docs, [README.md](../blob/main/README.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the demo app and try the changes? _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

ccjernigan commented on sdk-lib/src/main/java/cash/z/ecc/android/sdk/jni/RustBackend.kt:
```diff
@@ -56,18 +56,15 @@ class RustBackend private constructor() : RustBackendWelding {
-        val extfvks = Array(keys.size) { "" }
-        val extpubs = Array(keys.size) { "" }
+    override suspend fun initAccountsTable(vararg keys: UnifiedFullViewingKey): Boolean {
+        val ufvks = Array(keys.size) { "" }
```
A stylistic choice, but we can probably inline this now and drop the forEach below.  The lambda for the Array() constructor provides an index as `it` and the return value initializes that index.

`Array(keys.size) { keys[it].encoding }`

str4d replied:
I stuck close to the existing style because I'm not a Kotlin dev 😄 We should just clean this up across the board in a subsequent PR.

ccjernigan replied:
I've resolved this with an additional commit.

ccjernigan commented on sdk-lib/src/main/java/cash/z/ecc/android/sdk/type/WalletTypes.kt:
```diff
@@ -52,16 +52,16 @@ data class WalletBirthday(
-    val extfvk: String = "",
-    val extpub: String = ""
+data class UnifiedFullViewingKey(
+    val encoding: String = ""
```
This is a pre-existing issue.  I don’t think we ever want to allow `UnifiedFullViewingKey(“”)` 

Since we’re already making a breaking change here, can we drop the empty string as a default constructor parameter?

ccjernigan replied:
Also are there any invariants that you’d like to enforce here?  (Minimum length?)

ccjernigan replied:
I notice a security issue too.  We should probably drop `data` so this becomes an ordinary class. Data provides a default equals, hash code, and toString implementation.  The `toString` is dangerous in terms of accidentally logging sensitive information. (Alternative is to override the toString to block its default return value).

str4d replied:
This is actually something I want to overhaul before we make the breaking release. I agree with making this an opaque class, so we can better control how it is instantiated, and alter its internals freely without breaking changes. I'm not sure I want to do that in this PR though, which is focused on the UFVK breaking API changes that need to affect the FFI. I'll open an issue for addressing this.

str4d replied:
#531

str4d replied:
As we're now using a feature branch, we can ensure that we complete #531 before merging it.

ccjernigan commented on sdk-lib/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ edition = "2018"
 [dependencies]
 android_logger = "0.9"
 failure = "0.1"
+hdwallet = "=0.3.0"
```
I noticed most of our dependencies are not locked with `=`. Is there a particular reason this one is?

str4d replied:
The upstream developer merged one of our PRs to bump the `secp256k1` version from 0.19 to 0.21 (a breaking change), and released it as 0.3.1 (which Rust SemVer treats as backwards-compatible). We've migrated to `secp256k1 0.21` in `zcash/librustzcash`, but as of this PR we are still on `secp256k1 0.19` and thus need to pin this temporarily.

str4d replied:
This has since been fixed; we now depend on `hdwallet >= 0.3.1`.

ccjernigan reviewed (commented):
I am going to push commit to this branch later today which includes updates to the MIGRATIONS.md file.  Besides that needing to be updated, I think this looks good.

I did have a few very small comments/questions inline

nuttycom commented on sdk-lib/src/main/rust/utils.rs:
```diff
@@ -77,3 +89,75 @@ where
+}
+
+/// This is temporary, and will be replaced by `zcash_address::unified::Ufvk`.
+pub(crate) fn fake_ufvk_encode(p2pkh: &ExtendedPubKey, sapling: &ExtendedFullViewingKey) -> String {
```
Since zcash_address doesn't have any dependency upon zcash_primitives, why not just pull that in now? The `zcash_address` crate didn't even exist at the time of `zcash_primitives = "0.5"` so there isn't any conflict.

str4d replied:
I did try this, but none of the `zcash_primitives::zip32` refactors are present in this branch, so there's a mismatch between the types I have at my disposal, and the data that `zcash_address::unified` allows to be serialized. I figured this is only going to exist for a single PR, and then will disappear, so it's a helpful stepping stone.

str4d replied:
I was able to jump all the way to proper UFVK encodings in this PR, so this is now just an internal detail of the branch.

str4d commented on MIGRATIONS.md:
```diff
@@ -1,7 +1,24 @@
+For SDK clients that store the key separately from the mnemonic, the migration might look like:
+ * Replace usage of `UnifiedViewingKey(extfvk: String, extpub: String)` with `UnifiedFullViewingKey(encoding: String)`
+ * Replace `DerivationTool.deriveUnifiedViewingKeys` with `DerivationTool.deriveUnifiedFullViewingKeys`
+ * Delete any previously persisted values for `UnifiedViewingKey(extfvk: String, extpub: String)`.  The field `extpub` no longer exists.  The field `extfvk` needs to be re-generated from the mnemonic, and it has been renamed to `encoding` in `UnifiedFullViewingKey(encoding: String)`
```
This migration step is incorrect; the field `extfvk` technically doesn't exist any more either. Its value happens to be contained within the `encoding` field as part of the UFVK, but that is an internal detail of UFVKs that isn't relevant to the SDK user.

ccjernigan replied:
Feel free to propose alternative phrasing here or push another commit.

str4d replied:
Done.

str4d commented:
Rebased on current `main`.

str4d commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/da6ef7016e760a0c2b21dff220e1ec481939b9db..58641b7d12b3a0f64792fd2aa949f20b676a1460) to address my comment on `MIGRATIONS.md`.

pacu reviewed (commented):
no specific change. Please verify the changes on verified balance and total balance since regressions had been spotted here 
https://github.com/zcash/ZcashLightClientKit/issues/366

and here https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/8

str4d commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -467,6 +452,9 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_isValidTran
             Some(addr) => match addr {
                 RecipientAddress::Shielded(_) => Ok(JNI_FALSE),
                 RecipientAddress::Transparent(_) => Ok(JNI_TRUE),
+                // TODO: Should this count? What are the intended semantics of this FFI method?
```
Hmm, so the goal is to verify that it _is_ a t-addr, or is this used to show what kind of address has been scanned? If that's the case, then UAs should return `false` in both of these and we should have a separate API for exposing UA metadata. Or... I can just ignore that and wait until Uniffi gives us much nicer types that make these APIs obvious.

str4d replied:
@pacu and @ccjernigan confirmed to me that these APIs are used for _type_ checks (i.e. form validation), so my changes here were incorrect: I instead need to add an `isValidUnifiedAddress` FFI method. These will all be replaced by the `zcash_address` crate once we have Uniffi wrappers.

nuttycom commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -138,29 +140,22 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_initAccount
-                // TODO: replace with `zcash_address::unified::Ufvk`
-                utils::fake_ufvk_decode(&ufvkstr).ok_or_else(|| {
+                UnifiedFullViewingKey::decode(&network, &ufvkstr).map_err(|e| {
+                    if e.starts_with("UFVK is for network") {
```
This string matching is a bit ugly; can `UnifiedFullViewingKey::decode` return an enum for its error type instead?

str4d replied:
Yeah, we can rework this API once we've finished working out how it is used. I was more focused on other parts of the stack to get us to this point.

daira replied:
Please file an issue so that we don't lose track.

nuttycom commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -264,16 +258,15 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_tool_DerivationTool_deriveU
+            .map(|usk| usk.to_unified_full_viewing_key())?;
+
+        let (ua, _) = ufvk
+            .find_address(DiversifierIndex::new())
```
I think that this should be `DiversifierIndex::zero()` or `DiversifierIndex::from(0)`; `new` doesn't seem like the right constructor for `DiversifierIndex`.

str4d replied:
I prefer the `orchard` `DiversifierIndex` type, but we're in the `zcash_*` crates so I used the one from `zcash_primitives`. We should probably figure out a way to reify the types.

nuttycom replied:
Maybe `DiversifierIndex` should be in `zcash_address`?

nuttycom commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -292,22 +285,20 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_tool_DerivationTool_deriveU
-        let address_str = utils::fake_ua_encode(&p2pkh, &sapling);
+        // Derive the default Unified Address (containing the default Sapling payment
+        // address that older SDKs used).
+        let (ua, _) = ufvk.default_address();
```
This reminds me that we need to ensure that we handle search for the child 0 "diversified" transparent address correctly even when the Sapling default address is at a different diversifier index.

daira replied:
Please file an issue.

nuttycom replied:
https://github.com/zcash/librustzcash/issues/599

nuttycom commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -443,7 +426,9 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_isValidShie
             Some(addr) => match addr {
-                RecipientAddress::Shielded(_) => Ok(JNI_TRUE),
+                // TODO: What are the intended semantics of this FFI method?
+                // Unified Addresses always have a shielded component.
```
The SDKs will support sending to Sapling addresses in all cases, so this question just applies to receiving funds at bare Sapling addresses. I think that we still need to support this at the SDK level; users may need to share shielded addresses with people who haven't yet upgraded to a wallet with UA parsing support.

str4d replied:
This was clarified in a meeting: this API is for type checking (form validation). I've fixed it to treat `isValidShieldedAddress` as only checking for `RecipientAddress::Shielded`, with a new `isValidUnifiedAddress` handling `RecipientAddress::Unified`.

nuttycom reviewed (approved):
utACK with a minor comment about how the librustzcash API that this depends upon could be improved (hopefully before releasing this?)

ccjernigan reviewed (changes requested):
Besides a few small comments inline, I pushed some changes in separate commits to resolve build failures on CI (e.g. tests wouldn't compile, linting failed).

str4d commented on sdk-lib/src/main/java/cash/z/ecc/android/sdk/db/entity/Sent.kt:
```diff
@@ -26,6 +26,9 @@ data class Sent(
     val transactionId: Long = 0,
 
+    @ColumnInfo(name = "output_pool")
+    val outputPool: Int = 0,
```
I'd planned to handle migrations in a separate subsequent PR.

pacu replied:
Yes, we do. I had to do it in order for existing wallets to work

ccjernigan replied:
We try to have the main branch be always releasable, e.g. we don't want to intentionally merge a commit that we know will break the main branch for consumers of the SDK.

We're trying to resolve a few different bugs concurrently right now, so I think we need to keep the main branch stable.

We are also about as close to continuous deployment as we can be.  The Secant app consumes snapshots of the SDK so as soon as we merge this PR the Secant app will be ready to consume it.

We should address this migration in some way before merging this, even if we do a followup with an improved migration.

pacu replied:
I agree with carter. The migration follows the same pattern as existing ones.

str4d replied:
I somewhat disagree; my intention by splitting the PRs was to keep the process and review load manageable, as there was a _lot_ to do. But if reviewers are willing to review a noticeably more complex PR in once go, then we can combine them.

str4d replied:
In a meeting we decided that we'd merge this PR into a feature branch, so we could merge it without migrations (handling those in a separate PR), while waiting until the feature branch will not break `main` before merging it there.

pacu commented on sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/DerivationTool.kt:
```diff
@@ -87,17 +87,20 @@ class DerivationTool {
-        fun validateUnifiedViewingKey(viewingKey: UnifiedViewingKey, networkId: Int = ZcashNetwork.Mainnet.id) {
-            // TODO
-        }
+        // fun validateUnifiedFullViewingKey(
```
This is needed for clients implementing receive-only Point-Of-Sale solutions

str4d replied:
It's never been implemented, so *shrug*

In any case, the PR now keeps this method in (and it continues to be a no-op).

nuttycom commented:
Oops, I fouled up that rebase, going to redo it.

str4d commented:
[Rebased on `main`](https://github.com/zcash/zcash-android-wallet-sdk/compare/3218b18c91eafaf7c1b4c29b0990a6ff9843bf4f..ee8469d7948929fd8b5c5be054a48d3a24ffe41d) to fix merge conflicts due to the recent 1.8 release. I also cleaned up the commit history slightly, and re-targeted this PR to merge into the new `feature/zip-316-and-latest-upstream` branch, which we will use for staging purposes (so we don't break the `main` branch for existing wallets, which this PR alone will do).

str4d commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/ee8469d7948929fd8b5c5be054a48d3a24ffe41d..626f6cfa80a7b9cad1545e2e686621c82ca70831):
- Fixed TODOs related to my incomplete understanding of what `isValid{Shielded, Transparent}Address` were being used for; there is now a corresponding `isValidUnifiedAddress` and `AddressType.Unified`.
- Fixed the type signature of `Synchronizer.shieldFunds` (it compiled before because there's no actual type change, but one of the input variables was renamed in `SdkSynchronizer` but not here).
- Made a small `cargo fmt` change to the PR contents (the PR is not completely `cargo fmt` clean because the `main` branch is not; I've left that out of this PR).

str4d commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -520,7 +508,7 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_getVerified
 
         let amount = (&db_data)
-            .get_target_and_anchor_heights()
+            .get_target_and_anchor_heights(ANCHOR_OFFSET)
```
This change should be a no-op. Previously `get_target_and_anchor_heights` internally used a fixed offset, and this PR just maintains that fixed offset.

str4d commented on CHANGELOG.md:
```diff
@@ -12,6 +12,45 @@ Version 1.6.0-beta01
+  - `DerivationTool.deriveTransparentSecretKey` (use
+    `DerivationTool.deriveTransparentAccountPrivateKey` instead).
+  - `DerivationTool.deriveShieldedAddress`
+    - TODO: Do we still need to be able to derive Sapling shielded addresses for legacy
```
We can address this later on in the feature branch, once we've got a working integration with the Android apps.

str4d commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/626f6cfa80a7b9cad1545e2e686621c82ca70831..d7f779752e02be8df05c04874e0bc4e1c9cd647a) to undo some now-unnecessary compile warning fixes (that after rebasing on `main` began causing compile errors).

str4d commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/d7f779752e02be8df05c04874e0bc4e1c9cd647a..a86ba9855d7d548d641794bc4dab3110ad29089a) to squash one of the commits to clean up the PR history.

str4d commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/a86ba9855d7d548d641794bc4dab3110ad29089a..d1d406229880f0c12456ba87762454c13518cf91) to fix more Detekt warnings.

daira commented on MIGRATIONS.md:
```diff
@@ -1,7 +1,22 @@
+For SDK clients that store the key separately from the mnemonic, the migration might look like:
+ * Replace usage of `UnifiedViewingKey(extfvk: String, extpub: String)` with `UnifiedFullViewingKey(encoding: String)`
+ * Replace `DerivationTool.deriveUnifiedViewingKeys` with `DerivationTool.deriveUnifiedFullViewingKeys`
+ * Delete any previously persisted values for `UnifiedViewingKey(extfvk: String, extpub: String)`.
```
```suggestion
 * Delete any previously persisted values for `UnifiedViewingKey(extfvk: String, extpub: String)`,
   provided that they can be rederived from the mnemonic.
```

daira commented on darkside-test-lib/src/androidTest/java/cash/z/ecc/android/sdk/darkside/test/DarksideTestCoordinator.kt:
```diff
@@ -144,8 +144,8 @@ class DarksideTestCoordinator(val wallet: TestWallet) {
             runBlocking {
-                assertTrue((synchronizer as SdkSynchronizer).findBlockHashAsHex(height) != null)
-                assertTrue((synchronizer as SdkSynchronizer).findBlockHash(height)?.size ?: 0 > 0)
+                assertTrue((synchronizer).findBlockHashAsHex(height) != null)
```
Non-blocking nit:
```suggestion
                assertTrue(synchronizer.findBlockHashAsHex(height) != null)
```
and similarly in other cases.

daira commented on sdk-lib/Cargo.lock:
```diff
@@ -1598,53 +1904,77 @@ dependencies = [
+[[package]]
+name = "zcash_note_encryption"
+version = "0.1.0"
+source = "git+https://github.com/zcash/librustzcash.git?rev=b52e949bd6ee1dbbc0324437440bed8b3c423f79#b52e949bd6ee1dbbc0324437440bed8b3c423f79"
```
This is using two different revisions of zcash_note_encryption.

daira commented on sdk-lib/Cargo.toml:
```diff
@@ -12,28 +12,26 @@ edition = "2018"
+zcash_client_backend = { git = 'https://github.com/zcash/librustzcash.git', rev='b52e949bd6ee1dbbc0324437440bed8b3c423f79' }
+zcash_client_sqlite = { git = 'https://github.com/zcash/librustzcash.git', rev='b52e949bd6ee1dbbc0324437440bed8b3c423f79' }
+zcash_primitives = { git = 'https://github.com/zcash/librustzcash.git', rev='b52e949bd6ee1dbbc0324437440bed8b3c423f79' }
+zcash_proofs = { git = 'https://github.com/zcash/librustzcash.git', rev='b52e949bd6ee1dbbc0324437440bed8b3c423f79' }
```
Adding a patch for zcash_note_encryption here might fix the duplication, I think.

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -9,60 +9,61 @@ use std::str::FromStr;
 
 mod utils;
 
+const ANCHOR_OFFSET: u32 = 10;
```
This is not what we want long-term (see draft ZIP 315), but it will do for now.

str4d replied:
Yep, this is intentionally a no-op in this PR.

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -226,40 +219,30 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_tool_DerivationTool_deriveU
-            extpubs,
-            |env, extfvkstr| env.new_string(extfvkstr),
+            ufvks,
+            "java/lang/String",
```
Abstractions leaking, leaking everywhere. (Don't mind me. Just grumbling at the state of software infrastructure.)

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -444,7 +427,7 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_isValidShie
             Some(addr) => match addr {
                 RecipientAddress::Shielded(_) => Ok(JNI_TRUE),
-                RecipientAddress::Transparent(_) => Ok(JNI_FALSE),
+                RecipientAddress::Transparent(_) | RecipientAddress::Unified(_) => Ok(JNI_FALSE),
```
Document in the API docs that this returns false for unified addresses.

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -529,12 +534,13 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_getVerified
-            .sum::<Amount>();
+            .map(|utxo| utxo.txout.value)
+            .sum::<Option<Amount>>()
+            .ok_or_else(|| format_err!("Balance overflowed MAX_MONEY. Lucky you!"))?;
```
```suggestion
            .ok_or_else(|| format_err!("Balance overflowed MAX_MONEY"))?;
```

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -566,12 +572,13 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_getTotalTra
-            .sum::<Amount>();
+            .map(|utxo| utxo.txout.value)
+            .sum::<Option<Amount>>()
+            .ok_or_else(|| format_err!("Balance overflowed MAX_MONEY. Lucky you!"))?;
```
```suggestion
            .ok_or_else(|| format_err!("Balance overflowed MAX_MONEY"))?;
```

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -943,17 +969,31 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_tool_DerivationTool_deriveT
+        let index = if index >= 0 {
+            index as u32
+        } else {
+            return Err(format_err!("index argument must be positive"));
```
```suggestion
            return Err(format_err!("index argument must be nonnegative"));
```

pacu commented on sdk-lib/src/main/java/cash/z/ecc/android/sdk/db/entity/Account.kt:
```diff
@@ -11,11 +11,11 @@ data class Account(
 
     @ColumnInfo(name = "transparent_address")
-    val transparentAddress: String = ""
+    val transparentAddress: String? = ""
```
Do we still need this if we have the UFVK?

str4d replied:
No, but that requires a separate database change (specifically https://github.com/zcash/librustzcash/issues/489).

pacu reviewed (commented):
Looks good to me. 
Left a few unblocking comments.

Little bit jealous for this PR to be honest 😅

daira commented on sdk-lib/src/main/rust/lib.rs:
```diff
@@ -1062,7 +1109,7 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_jni_RustBackend_createToAdd
 
```
On line 1078, "account argument must be nonnegative". Also look for other incorrect uses of "positive".

daira commented on sdk-lib/src/main/rust/utils.rs:
```diff
@@ -41,39 +51,63 @@ where
-            *jni::objects::JObject::null(),
-        )
-        .expect("Couldn't create Java array of string arrays!");
+//pub(crate) fn rust_vec_to_java_2d<'a, T, V, F, G>(
```
Just delete this. It can always be restored if needed.

daira reviewed (approved):
utACK modulo suggestions.

nuttycom reviewed (approved):
utACK for merge to the feature branch. I will address @daira's documentation comments and then merge.

nuttycom commented:
[Force-pushed](https://github.com/zcash/zcash-android-wallet-sdk/compare/d1d406229880f0c12456ba87762454c13518cf91..a3bb9cb7bb86ab3fc1f507f5513ea89c2b7fc173) to address review comments.

nuttycom reviewed (approved):
re-utACK a3bb9cb7bb86ab3fc1f507f5513ea89c2b7fc173

nuttycom reviewed (approved):
re-utACK https://github.com/zcash/zcash-android-wallet-sdk/pull/530/commits/7152c7be431bacd5602030c4fc4c29a4d9a1cd82
